scanner/enricher/csaf/parser.go

```diff
@@ -0,0 +1,105 @@
+package csaf
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/klauspost/compress/snappy"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/toolkit/types/csaf"
+	"github.com/quay/zlog"
+)
+
+// ParseEnrichment implements driver.EnrichmentUpdater.
+func (e *Enricher) ParseEnrichment(ctx context.Context, contents io.ReadCloser) ([]driver.EnrichmentRecord, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "enricher/csaf/Enricher/ParseEnrichment")
+
+	records := make(map[string]Record)
+
+	r := bufio.NewReader(snappy.NewReader(contents))
+	for b, err := r.ReadBytes('\n'); err == nil; b, err = r.ReadBytes('\n') {
+		c, err := csaf.Parse(bytes.NewReader(b))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing CSAF: %w", err)
+		}
+		if c.Document.Tracking.Status == "deleted" {
+			continue
+		}
+
+		record := Record{
+			// This would be the RHSA/RHBA/RHEA ID.
+			Name: c.Document.Tracking.ID,
+			// Use the title as the advisory's description.
+			// The current csaf type from Claircore as of toolkit/v1.2.4 does not enable us
+			// to fetch c.Document.Notes which would have more in-depth descriptions.
+			// The title is a decent compromise for now.
+			Description: c.Document.Title,
+			// The initial_release_date is the date when the advisory was published.
+			// The current_release_date is the last-updated date, so
+			// we use initial_release_date here.
+			ReleaseDate: c.Document.Tracking.InitialReleaseDate,
+			// Obtain the aggregate severity rating of all related CVEs for this advisory,
+			// which tends to be the highest severity of all related CVEs.
+			// This matches the severity we'd obtain from OVAL.
+			Severity: c.Document.AggregateSeverity.Text,
+		}
+
+		// Back in the Red Hat OVAL days, we would assign an advisory the highest-related CVSS scores.
+		// We mimic that functionality here.
+		//
+		// Note: it is very possible a CVE has two different CVSS scores, depending on the product.
+		// For example: https://access.redhat.com/security/cve/CVE-2023-3899 is scored 7.8, in general,
+		// but 6.1 for subscription-manager in RHEL 7.
+		// For this case, the OVAL v2 entry in https://security.access.redhat.com/data/oval/v2/RHEL7/rhel-7-including-unpatched.oval.xml.bz2
+		// for the associated RHSA, RHSA-2023:4701, actually has the general CVSS score (7.8) instead of the true score (6.1).
+		// Meanwhile, the CSAF entry in https://security.access.redhat.com/data/csaf/v2/advisories/2023/rhsa-2023_4701.json
+		// lists the true, correct score of 6.1.
+		// So, the output we get here will differ from the previous OVAL v2-based output, but it will be more accurate
+		// (though we acknowledge Red Hat advisories should not really be assigned CVSS scores in the first place).
+		var cvss2, cvss3 struct {
+			score  float64
+			vector string
+		}
+		for _, v := range c.Vulnerabilities {
```

Comment:
After discussing with @jvdm, I noticed that while we are already obtaining the highest CVSS score among all vulnerabilities, I’m wondering if we could also retrieve the list of CVE names for an RHSA or the specific CVE with the highest CVSS score. This would allow us to either use the CVE name list to determine the highest EPSS probability or directly fetch the EPSS probability for the CVE with the highest CVSS score. Either approach would help ensure consistency in the EPSS score. Please let me know your thoughts.

Reply:
Related conversation on the EPSS ADR: https://github.com/stackrox/stackrox/pull/13752/files#r1919174380

Reply:
I'll add something to track all the names of the related CVEs, so we may choose the highest EPSS score from them

Reply:
Noted elsewhere but will put here for completeness: we found that doing this is non-trivial, so I won't track the "sub CVEs" for an RHSA. I can update the EPSS ADR in a followup PR